src/lib/cache.coffee

```diff
@@ -0,0 +1,83 @@
+###
+
+dnschain
+http://dnschain.net
+
+Copyright (c) 2014 okTurtles Foundation
+
+This Source Code Form is subject to the terms of the Mozilla Public
+License, v. 2.0. If a copy of the MPL was not distributed with this
+file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+###
+
+redis = require 'redis'
+
+module.exports = (dnschain) ->
+    # expose these into our namespace
+    for k of dnschain.globals
+        eval "var #{k} = dnschain.globals.#{k};"
+
+    class ResolverCache
+        constructor: (@dnschain) ->
+            @log = gNewLogger 'CACHE'
+            @log.info "Loading Redis Cache..."
+            if gConf.get('redis:blockchain:enabled') or gConf.get('redis:oldDNS:enabled')
+                @log.info "cache enabled"
+                [host,port] = gConf.get('redis:socket').split(':')
+                @cache = if port?
+                    port = parseInt port
+                    redis.createClient(port, host)
+                else
+                    redis.createClient(host)
+                if gConf.get('redis:blockchain:enabled')
+                    @blockchainEnabled = true
+                if gConf.get('redis:oldDNS:enabled')
+                    @oldDNSEnabled = true
+                    @oldDNSTTL = 600
+                @cache.on 'error', (err) =>
+                    @log.error "cache errored"
+                    @shutdown()
+
+            else
+                @log.info "cache disabled"
+
+        resolveBlockchain: (resolver, path, options, cb) ->
+            if @blockchainEnabled? and resolver.cacheTTL?
+                @cache.get "#{resolver.name}:#{path}:#{JSON.stringify(options)}", (err, result) =>
```

Comment:
Yeah, I agree there is some code repetition going on between these functions in `ResolverCache` and the functions that are using them (in `dns.coffee`, `http.coffee`, etc.).

The primary problem is the arguments you are passing to this function and what you are choosing to do with them.

You have two options, either:

1. Move all the logic stuff that's currently in `dns.coffee` into here
2. Or, make this function more generic and have it act on key/value pairs only, in which case, you shouldn't be constructing the key string here, but outside of this function, and then if you don't get a cache hit, you call `resolver.resolve` by calling a callback function that's passed in to this function.